luigi_pipeline/lib/model/base_mt_schema.py

```diff
@@ -0,0 +1,104 @@
+from abc import abstractmethod
+from inspect import getmembers, isfunction
+from functools import wraps
+
+
+def mt_annotation(annotation=None, fn_require=None):
+    """
+    Function decorator for methods in a subclass of BaseMTSchema.
+    Allows the function to be treated like an mt_annotation with annotation name and value.
+
+        @mt_annotation()
+        def a(self):
+            return 'a_val'
+
+        @mt_annotation(annotation='b', fn_require='a')
+        def b_1(self):
+            return 'b_val'
+
+    Will generate a mt with rows of {a: 'a_val', 'b': 'b_val'} if the function is called.
+
+    :param annotation: name in the final MT. If not provided, uses the function name.
+    :param fn_require: method name strings in class that are dependencies.
+    :return:
+    """
+    def mt_prop_wrapper(func):
+        annotation_name = annotation or func.__name__
+
+        @wraps(func)
+        def wrapper(self, *args, **kwargs):
+            # Called already.
+            if wrapper.mt_prop_meta['annotated'] > 0:
+                return self
+            if fn_require:
+                getattr(self, fn_require)()
+
+            # Annotate and retiurn instance for chaining.
+            self.mt = self.mt.annotate_rows(**{annotation_name: func(self, *args, **kwargs)})
+            wrapper.mt_prop_meta['annotated'] += 1
+            return self
+
+        wrapper.mt_prop_meta = {
+            'annotated': 0,  # Counter for number of times called. Should only be 0 or 1.
```

Comment:
This is for bookkeeping (helpful in our tests to guarantee that), but in the future something could be added to log a warning if it goes over 1 (didn't add it yet because I didn't think it was needed).